FEATURE: Speedup content cache flush by using cte in `findAncestorNodeAggregateIds`

mhsdesign opened the pull request:
<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

Followup to #5221


`ContentGraph::findParentNodeAggregates` becomes slower on bigger datasets. Due to the mass on executions on a `cr:replay`, this sums up very quickly. Via https://github.com/neos/neos-development-collection/pull/5268 the query will be improved but this pr introduces `ContentGraph::findAncestorNodeAggregateIds` to make this operation as performant as possible.


- Adds comment for `CacheFlushingStrategy` strategies
- Introduces `ContentGraphInterface::findAncestorNodeAggregateIds` using native sql cte to speedup cache flushing. (see [comment](https://github.com/neos/neos-development-collection/pull/5221#discussion_r1742654587))
- Move test which creates illegal state to content graph package and use native sql to create the state to not run any catchup hooks. Previously we needed to handle the case of infinite loops to not crash:

  > Prevent infinite loops
  > NOTE: Normally, the content graph cannot contain cycles. However, during the
  > testcase "Features/ProjectionIntegrityViolationDetection/AllNodesAreConnectedToARootNodePerSubgraph.feature"
  > and in case of bugs, it could have actually cycles.
  > The content cache catchup hook leverage this method and would otherwise be hanging up in an endless loop.
  > That's why we track the seen NodeAggregateIds to be sure we don't travers them multiple times.

- fixes a bug where you cannot replay because the workspace is "missing" and no content graph exists


**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Add docs to `CacheFlushingStrategy`

mhsdesign committed:
TASK: introduce `ContentGraphInterface::findAncestorNodeAggregateIds`

mhsdesign committed:
WIP: remove hacky loop detection in `findAncestorNodeAggregateIds` by applying event in test directly

mhsdesign commented on Neos.ContentRepository.BehavioralTests/Tests/Behavior/Features/ProjectionIntegrityViolationDetection/AllNodesAreConnectedToARootNodePerSubgraph.feature:
```diff
@@ -52,7 +52,8 @@ Feature: Run projection integrity violation detection regarding root connection
       | nodeName                    | "child-document"                          |
       | nodeAggregateClassification | "regular"                                 |
-    And the event NodeAggregateWasMoved was published with payload:
+    # we must apply the event directly as the state is so corrupt that catchup hooks might fail like the content cache flusher
+    And the event NodeAggregateWasMoved is hacky directly applied with payload:
```
this is just a random idea in the totally optional last commit that we can drop as well. The idea is that we dont even really publish the event but call `DoctrineDbalContentGraphProjection::apply` which is of course super hacky and i know but it will make sure no catchup hooks are triggered.

The alternative is to create the state by sql by hand and thats a mess probably?

mhsdesign replied:
Moved the test to the doctrine dbal and using now raw sql to create the invalid state.

mhsdesign committed:
TASK: Remove apply event hack and change hierarchy relation's parent manually

mhsdesign committed:
TASK: Prefer commands instead publishing events directly

dlubitz commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php:
```diff
@@ -188,6 +189,19 @@ public function findParentNodeAggregates(
         return $this->mapQueryBuilderToNodeAggregates($queryBuilder);
     }
 
+    public function findAncestorNodeAggregateIds(NodeAggregateId $entryNodeAggregateId): NodeAggregateIds
```
As far a I understood, this function needs to be more complex and return a set of pathways with ancestor nodes.
https://neos-project.slack.com/archives/C04PYL8H3/p1725605274115779?thread_ts=1725460465.658809&cid=C04PYL8H3

mhsdesign replied:
Hmm but to get that right, does that mean that if implemented more complex also other node ids will be returned or just that all the node ids will be nicely ordered by dimension in a special value object. The latter i consider just an improvement and not critical. The idea is to improve the performance of the method via native sql at first. Its marked as `@internal` so we can improve the structure later?

dlubitz replied:
> the node ids will be nicely ordered by dimension in a special value object

Yes this one

> Hmm but to get that right, does that mean that if implemented more complex also other node ids will be returned or just that all the node ids will be nicely ordered by dimension in a special value object. The latter i consider just an improvement and not critical. The idea is to improve the performance of the method via native sql at first. Its marked as @internal so we can improve the structure later?

But you don't improve that with this PR. 
And even if it is marked as internal, i think we shouldn't put that into an interface  if we already know, that we will change that.

mhsdesign replied:
> But you don't improve that with this PR.

i _could_ try to do that ... emphasis on try :D or ask bernhard. But i get your point. Lets see how we can continue.

cc @nezaniel @bwaidelich @kitsunet 

edit: and we def need behat tests for the method i forgot :D

mhsdesign replied:
added behat tests and introduced cte query to improve performance. Its now ordered desc by the parentnodeanchor

mhsdesign committed:
TASK: Add behat test for `findAncestorNodeAggregateIds`

mhsdesign committed:
TASK: Add behat test for multiple parent node aggregates

mhsdesign committed:
FEATURE: Optimise `findAncestorNodeAggregateIds` to use CTE and sort result

mhsdesign committed:
WIP position ordering for sibling in `findAncestorNodeAggregateIds`

mhsdesign committed:
TASK: Simplify code in `GraphProjectorCatchUpHookForCacheFlushing`

mhsdesign committed:
TASK: Do not sort `findAncestorNodeAggregateIds`

In the case of the content cache flusher we do not care about the order and ordering it by parentnodeanchor and position (for siblings) is slower and not even correct in all situations as the parentnodeanchor is just an autoincrement without meaning.

mhsdesign committed:
BUGFIX: Cannot replay "The source workspace missing does not exist"

dlubitz commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php:
```diff
@@ -188,6 +189,42 @@ public function findParentNodeAggregates(
+            ->select('n.nodeAggregateId, ph.parentnodeanchor')
+            ->from($this->nodeQueryBuilder->tableNames->node(), 'n')
+            // we need to join with the hierarchy relation, because we need the node name.
+            ->innerJoin('n', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ch', 'ch.parentnodeanchor = n.relationanchorpoint')
+            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'c', 'c.relationanchorpoint = ch.childnodeanchor')
+            ->innerJoin('n', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ph', 'n.relationanchorpoint = ph.childnodeanchor')
```
This query again joins, the hierachy relation twice. I think this is not needed. See: https://github.com/neos/neos-development-collection/pull/5268

nezaniel replied:
has been adjusted

mhsdesign committed:
TASK: Remove obsolete join

Commit:
Merge branch '9.0' into task/contentCacheFlusher-followup

nezaniel commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php:
```diff
@@ -187,6 +188,39 @@ public function findParentNodeAggregates(
+        $queryBuilderRecursive = $this->createQueryBuilder()
+            ->select('pn.nodeAggregateId, h.parentnodeanchor')
+            ->from('ancestry', 'cn')
+            ->innerJoin('cn', $this->nodeQueryBuilder->tableNames->node(), 'pn', 'pn.relationanchorpoint = cn.parentnodeanchor')
```
cn should actually be named ch, right? It's joined as the child hierarchy relation
h could then be ph, just for better understanding

nezaniel reviewed (dismissed):
minimal naming stuff, looks fine otherwise (and is a lot faster in the non-trivial test cases)

mhsdesign committed:
TASK: Task rename alias in ancestor queries

`cn` should actually be named `ch` as it's joined as the child hierarchy relation

dlubitz commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentGraph.php:
```diff
@@ -187,6 +189,39 @@ public function findParentNodeAggregates(
 
+    public function findAncestorNodeAggregateIds(NodeAggregateId $entryNodeAggregateId): NodeAggregateIds
+    {
+        $queryBuilderInitial = $this->createQueryBuilder()
+            ->select('n.nodeAggregateId, ch.parentnodeanchor')
+            ->from($this->nodeQueryBuilder->tableNames->node(), 'n')
+            ->innerJoin('n', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ch', 'ch.parentnodeanchor = n.relationanchorpoint')
+            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'c', 'c.relationanchorpoint = ch.childnodeanchor')
+            ->where('ch.contentstreamid = :contentStreamId')
+            ->andWhere('c.nodeaggregateid = :entryNodeAggregateId');
+
+        $queryBuilderRecursive = $this->createQueryBuilder()
+            ->select('pn.nodeAggregateId, ph.parentnodeanchor')
+            ->from('ancestry', 'ch')
+            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'pn', 'pn.relationanchorpoint = ch.parentnodeanchor')
+            ->innerJoin('pn', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ph', 'ph.childnodeanchor = pn.relationanchorpoint')
+            ->where('ph.contentstreamid = :contentStreamId');
+
+        $queryBuilderCte = $this->createQueryBuilder()
+            ->select('pn.nodeAggregateId')
+            ->from('ancestry', 'pn')
+            ->setParameter('contentStreamId', $this->contentStreamId->value)
+            ->setParameter('entryNodeAggregateId', $entryNodeAggregateId->value);
```
```suggestion
        $queryBuilderInitial = $this->createQueryBuilder()
            ->select('ch.parentnodeanchor')
            ->from($this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ch')
            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'c', 'c.relationanchorpoint = ch.childnodeanchor')
            ->where('ch.contentstreamid = :contentStreamId')
            ->andWhere('c.nodeaggregateid = :entryNodeAggregateId');

        $queryBuilderRecursive = $this->createQueryBuilder()
            ->select('ph.parentnodeanchor')
            ->from('ancestry', 'ch')
            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->hierarchyRelation(), 'ph', 'ph.childnodeanchor = ch.parentnodeanchor')
            ->where('ph.contentstreamid = :contentStreamId');

        $queryBuilderCte = $this->createQueryBuilder()
            ->select('n.nodeAggregateId')
            ->from('ancestry', 'a')
            ->innerJoin('a', $this->nodeQueryBuilder->tableNames->node(), 'n', 'n.relationanchorpoint = a.parentnodeanchor')
            ->setParameter('contentStreamId', $this->contentStreamId->value)
            ->setParameter('entryNodeAggregateId', $entryNodeAggregateId->value);
```

The query could be optimized a bit more (at least for deeper trees), if we join the node table only after we built the hierarchy.

mhsdesign replied:
thanks ❤️ applied that:) idk understand it yet exactly but it works ^^

dlubitz commented on Neos.ContentGraph.DoctrineDbalAdapter/src/Domain/Repository/ContentSubgraph.php:
```diff
@@ -367,8 +368,8 @@ public function findClosestNode(NodeAggregateId $entryNodeAggregateId, FindClose
             ->select('pn.*, h.subtreetags, h.parentnodeanchor')
-            ->from('ancestry', 'cn')
-            ->innerJoin('cn', $this->nodeQueryBuilder->tableNames->node(), 'pn', 'pn.relationanchorpoint = cn.parentnodeanchor')
+            ->from('ancestry', 'ch')
+            ->innerJoin('ch', $this->nodeQueryBuilder->tableNames->node(), 'pn', 'pn.relationanchorpoint = ch.parentnodeanchor')
```
I'm not sure if this change is correct. Actually we get a node with tags and parentnodeanchor.
From my POV it's rather a node than a hierarchy. BUT ... 

Moreover I don't understand why this needs to be in this PR, as this is about "findAncestorNodeAggregateIds"

(same for the other places)

mhsdesign replied:
i it was from the review comment here https://github.com/neos/neos-development-collection/pull/5261#discussion_r1813588639 and as i copied the query mostly it had to be also "wrong" in other places ...

mhsdesign replied:
reverted that line:)

dlubitz commented:
Looks fine to me by reading. I left two comments.

mhsdesign committed:
Merge remote-tracking branch 'origin/9.0' into task/contentCacheFlusher-followup

mhsdesign committed:
TASK: Optimise `findAncestorNodeAggregateIds`

The query could be optimized a bit more (at least for deeper trees), if we join the node table only after we built the hierarchy.

Co-authored-by: Denny Lubitz <lubitz@vivomedia.de>

bwaidelich reviewed (approved):
+1 by reading, thanks!
